service/matching/matchingEngine.go

```diff
@@ -1180,11 +1178,21 @@ func (e *matchingEngineImpl) redirectToVersionedQueueForAdd(
 	taskQueue *taskQueueID,
 	stamp *commonpb.WorkerVersionStamp,
 	kind enumspb.TaskQueueKind,
+	isFirstWorkflowTask bool,
 ) (*taskQueueID, error) {
 	// sticky queues are unversioned
 	if kind == enumspb.TASK_QUEUE_KIND_STICKY {
 		return taskQueue, nil
 	}
+
+	// If this is the first workflow task for a workflow, we need to assign it to the default
+	// version. If not, we leave it on the unversioned queue, even if the queue has versioning
+	// data. We can check this before loading versioning data.
+	if stamp == nil && !isFirstWorkflowTask {
```

Comment:
See the condition here (requires updated protos though so not blocking your PR)
https://github.com/temporalio/temporal/pull/4284/files#diff-2be560efd088cb934728b0ebe0263e3a558468b43b5de8132fbf3e13942cd049R1188

Reply:
I'd prefer if we do the check in history and not send a stamp at all to matching

Reply:
Then I'd send a different message and not the stamp because the `use_versioning` flag is meaningless on the receiver side which I find confusing.

Reply:
> I mean, isn't that confusion exactly why I was arguing for keeping stamp a stamp instead of adding other semantic stuff to it?

I was arguing the same thing but we said this was the simplest way forward...



> Anyway, it's not clear what you're arguing for here. You want to not send WorkerVersioningStamp to matching at all and just pull out the build id??

My alternative would be to put more logic in history and send something like: `unversioned | unset | versioned(buildId)` to matching. `unversioned` would have to be the default because of existing / spooled tasks.

Reply:
Hmm.. actually, spooled tasks could be either unset or unversioned.

Reply:
But matching will have to assume unversioned.

Reply:
Existing from before this code is deployed? Yes, I was assuming those will stay on the unversioned queue (since in this iteration of the PR, they're missing is_first_workflow_task).

I definitely want to put more logic in history, it's just what it should look like... the tri-state feels like overkill but I'm coming around to the idea that it's the cleanest.

Reply:
Yeah, it seems like it captures the intention well and will be easy to extend to handle the activity/child workflow/c-a-n flag